reasoner/resolution/answer/AnswerState.java

```diff
@@ -0,0 +1,235 @@
+/*
+ * Copyright (C) 2020 Grakn Labs
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU Affero General Public License as
+ * published by the Free Software Foundation, either version 3 of the
+ * License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU Affero General Public License for more details.
+ *
+ * You should have received a copy of the GNU Affero General Public License
+ * along with this program.  If not, see <https://www.gnu.org/licenses/>.
+ */
+
+package grakn.core.reasoner.resolution.answer;
+
+import grakn.core.common.exception.GraknException;
+import grakn.core.concept.Concept;
+import grakn.core.concept.answer.ConceptMap;
+import graql.lang.pattern.variable.Reference;
+
+import javax.annotation.Nullable;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.Objects;
+import java.util.Optional;
+
+import static grakn.common.util.Objects.className;
+import static grakn.core.common.exception.ErrorMessage.Internal.ILLEGAL_STATE;
+import static grakn.core.common.exception.ErrorMessage.Pattern.INVALID_CASTING;
+
+public abstract class AnswerState {
```

Comment:
looking better! May i suggest the following naming change to see if anything falls out of it:
`UpstreamVars` -> `RequesterVars`
`DownstreamVars` -> `ResponderVars`

i think "upstream/downstream" (at least for me) is too closely tied to the shape of the actor execution, whereas "request/responder" is more generic? then it can read as `AnswerState.RequesterVars.Partial`, do you like that better?

should also replace occurrences of "Map<Ref, Set<Ref>>` with Unifier now i think

Reply:
Totally open to re-hashing the names. Have gone with upstream/downstream as that's made things very clear in the implementation. Currently the usage goes: 
`fromUpstream.partial().aggregateWith(conceptMap).asMapped().toUpstreamVars()`
with this proposal it would be:
`fromUpstream.partial().aggregateWith(conceptMap).asMapped().toRequesterVars()`
what I like about the former is that it's easy to tell that it's the variables of the "parent" actor, whereas in the latter this actor is also a requester. DOn't know if this argument really holds or if I just like the upstream/downstream variable naming 🤔

Reply:
Yes, can introduce `Unifier` as a data structure.

Reply:
didn't you arleady do that`?

Reply:
ok lets stick with your names for now :)

Reply:
Here actually I call the transformer that applies the mapping a `Unifier` - so we should rename this to... a `Unificator`? 🤣 
But really, what shall we call these two things?

Reply:
Let's get this merged and address the unifier naming when we implement the logic for unifier variable mapping and filtering.